### PR TITLE
Upgrade mechanism up to otelcol 0.10.0

### PR DESCRIPTION
**Pull request by jpkrohling**
- Improved upgrade mechanism
- Added migration to v0.9.0 (v0.10.0 was noop)

Signed-off-by: Juraci Paixão Kröhling juraci@kroehling.de

### CODE REVIEW

File: pkg/collector/upgrade/v0_9_0.go [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+	if !ok {
+		return otelcol, fmt.Errorf("couldn't upgrade to v0.9.0, failed to extract list of exporters from the configuration: %q", cfg["exporters"])
+	}
+
+	for k, v := range exporters {
+		if strings.HasPrefix("opencensus", k.(string)) {
+			switch exporter := v.(type) {
+			case map[interface{}]interface{}:
+				// delete is a noop if there's no such entry
+				delete(exporter, "reconnection_delay")
```
Comment (jpkrohling) [inline pkg/collector/upgrade/v0_9_0.go:34]: As discussed, we should also add a note to the `.Status` object.

File: pkg/collector/upgrade/upgrade.go [hunk 2]
```diff
@@ -59,20 +61,36 @@ func ManagedInstances(ctx context.Context, logger logr.Logger, cl client.Client)
...
+	}
+
+	instanceV, err := semver.NewVersion(otelcol.Status.Version)
+	if err != nil {
+		logger.Error(err, "failed to parse version for OpenTelemetry Collector instance", "name", otelcol.Name, "namespace", otelcol.Namespace, "version", otelcol.Status.Version)
+		return otelcol, err
+	}
+
+	if instanceV.GreaterThan(&Latest.Version) {
+		logger.Info("skipping upgrade for OpenTelemetry Collector instance, as it's newer than our latest version", "name", otelcol.Name, "namespace", otelcol.Namespace, "version", otelcol.Status.Version)
```
Comment (objectiser) [inline pkg/collector/upgrade/upgrade.go:78]: Might be good to include the latest version in the info message aswell?

File: pkg/collector/upgrade/upgrade.go [hunk 3]
```diff
@@ -59,20 +61,36 @@ func ManagedInstances(ctx context.Context, logger logr.Logger, cl client.Client)
...
 			}
 
-			logger.V(1).Info("step upgrade", "name", otelcol.Name, "namespace", otelcol.Namespace, "version", n.Tag)
-			upgraded.Status.Version = n.Tag
-			otelcol = upgraded
+			logger.V(1).Info("step upgrade", "name", otelcol.Name, "namespace", otelcol.Namespace, "version", available.String())
+			upgraded.Status.Version = available.String()
+			otelcol = *upgraded
 		}
 	}
```
Comment (objectiser) [inline pkg/collector/upgrade/upgrade.go:95]: As the loop only includes versions that have an associated upgrade function, we still need to set the `upgraded.Status.Version` to `currentV.OpenTelemetryCollector` before exiting?

File: pkg/collector/upgrade/versions.go [hunk 4]
```diff
@@ -1,35 +1,31 @@
...
+			upgrade: upgrade0_9_0,
+		},
 	}
-)
 
-func noop(cl client.Client, otelcol *v1alpha1.OpenTelemetryCollector) (*v1alpha1.OpenTelemetryCollector, error) {
-	return otelcol, nil
-}
+	// Latest represents the latest version that we need to upgrade. This is not necessarily the latest known version.
+	Latest = versions[len(versions)-1]
```
Comment (objectiser) [inline pkg/collector/upgrade/versions.go:30]: Just wondering if this is necessary? i.e. what purpose does it serve. The important versions are the instance's current version, and the max version supported by the operator?

File: config/manager/kustomization.yaml [hunk 5]
```diff
@@ -1,2 +1,8 @@
 resources:
 - manager.yaml
+apiVersion: kustomize.config.k8s.io/v1beta1
+kind: Kustomization
+images:
+- name: controller
+  newName: quay.io/jpkroehling/opentelemetry-operator
```
[the remark this answers is not in the document]
Comment (jpkrohling) [reply config/manager/kustomization.yaml:7]: This change keeps coming in... I have to figure out which part of make manifests is creating this...

File: config/samples/core_v1alpha1_opentelemetrycollector.yaml [hunk 6]
```diff
@@ -3,6 +3,7 @@ kind: OpenTelemetryCollector
 metadata:
   name: opentelemetrycollector-sample
 spec:
+  mode: deployment
```
[the remark this answers is not in the document]
Comment (jpkrohling) [reply config/samples/core_v1alpha1_opentelemetrycollector.yaml:6]: It is the default, and the Defaulter webhook will set it to `deployment` if empty, but when doing local development, this field will not be set when the reconciliation first kicks in, meaning that no deployments nor daemonsets will be created. I'll revert this change, and probably add a note in the readme/contributing about it.

File: pkg/collector/reconcile/configmap.go [hunk 7]
```diff
@@ -79,6 +79,13 @@ func expectedConfigMaps(ctx context.Context, params Params, expected []corev1.Co
 		err := params.Client.Get(ctx, nns, existing)
 		if err != nil && errors.IsNotFound(err) {
 			if err := params.Client.Create(ctx, &desired); err != nil {
+				if errors.IsAlreadyExists(err) {
+					// let's try again? we probably had multiple updates at one, and now it exists already
+					if err := expectedConfigMaps(ctx, params, expected, false); err != nil {
+						// somethin else happened now...
+						return err
+					}
+				}
```
[the remark this answers is not in the document]
Comment (jpkrohling) [reply pkg/collector/reconcile/configmap.go:88]: Great catch, will fix it.

File: pkg/collector/reconcile/opentelemetry.go [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+	changed := params.Instance
+
+	labels := changed.Labels
+	if labels == nil {
+		labels = map[string]string{}
+	}
+
+	// if it's not empty, and it's not opentelemetry-operator, then it's set to something else,
+	// and we don't want to change it
+	if labels["app.kubernetes.io/managed-by"] != "" && labels["app.kubernetes.io/managed-by"] != "opentelemetry-operator" {
```
[the remark this answers is not in the document]
Comment (jpkrohling) [reply pkg/collector/reconcile/opentelemetry.go:23]: When the CR is first created, this label isn't there. I guess I should actually move this to the defaulter too.

File: pkg/collector/upgrade/versions.go (hunk 4 above)
Comment (jpkrohling) [reply pkg/collector/upgrade/versions.go:30]: We have a special condition in the logic where we would skip the upgrade if the version set in the CR is higher than the one we known, like in the case for downgrades. If I expose the whole map of versions, I can remove this latest, and calculate the latest on those two places.

File: pkg/collector/reconcile/opentelemetry.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (objectiser) [reply pkg/collector/reconcile/opentelemetry.go:23]: +1

File: config/manager/kustomization.yaml (hunk 5 above)
[the remark this answers is not in the document]
Comment (jpkrohling) [reply config/manager/kustomization.yaml:7]: Found out what causes it:

```Makefile
# Deploy controller in the configured Kubernetes cluster in ~/.kube/config
deploy: manifests kustomize
	cd config/manager && $(KUSTOMIZE) edit set image controller=${IMG}
	$(KUSTOMIZE) build config/default | kubectl apply -f -
```

Not sure what to do here, but certainly not something for this PR :-)

**Comment by jpkrohling** [comment]
Most of the commends were addressed, the following is still pending:

- [ ] add a list of removed fields to the status object
- [ ] `Latest` vs. `Versions` (as opposed to `versions`)

**Comment by jpkrohling** [comment]
The Status object now looks like this when a field is removed:

```yaml
spec:
  config: |
    exporters:
      opencensus:
        endpoint: 0.0.0.0:55678
        num_workers: 123
    processors:
      queued_retry: null
    receivers:
      jaeger:
        protocols:
          grpc: null
    service:
      pipelines:
        traces:
          exporters:
          - opencensus
          processors:
          - queued_retry
          receivers:
          - jaeger
  mode: deployment
status:
  messages:
  - upgrade to v0.9.0 removed the property reconnection_delay for exporter "opencensus"
  version: 0.10.0
```

**Comment by jpkrohling** [comment]
@objectiser is this OK to be merged?

**Review by objectiser — APPROVED** [review]

File: api/v1alpha1/opentelemetrycollector_types.go [hunk 9]
```diff
@@ -74,11 +74,14 @@ type OpenTelemetryCollectorSpec struct {
 
 // OpenTelemetryCollectorStatus defines the observed state of OpenTelemetryCollector
 type OpenTelemetryCollectorStatus struct {
+	// Version of the managed OpenTelemetry Collector (operand)
 	// +optional
-	Replicas int32 `json:"replicas"`
```
[the remark this answers is not in the document]
Comment (jpkrohling) [reply api/v1alpha1/opentelemetrycollector_types.go]: I'll remove this, as it doesn't belong to this PR. The reason to remove it is that we might want to think again what "replicas" mean, especially when talking about `mode: sidecar`. Also, do we count all currently running replicas, or the desired ones?

**Pull request merged**